core/data_stream.py

```diff
@@ -0,0 +1,249 @@
+import abc
+import tempfile
+import os
+import typing
+from core.data_block import DataBlock
+
+Symbol = typing.Any
+
+
+class DataStream(abc.ABC):
+    """abstract class to represent a Data Stream
+
+    The DataStream facilitates the block interface.
+    From the interface standpoint, the two functions which are useful are:
+    - get_block(block_size) -> returns a DataBlock of the given block_size from the stream
+    - write_block(block) -> writes the block of data to the stream
+
+    The DataStream can act as a stream object for both writing and reading blocks
+    The two more useful sub-classes of the abstract class are FileDataStream and ListDataStream.
+    (see their description for more details)
+    """
+
+    @abc.abstractmethod
+    def reset(self):
+        """resets the data stream"""
+        pass
+
+    @abc.abstractmethod
+    def get_symbol(self):
+        """returns a symbol from the data stream, returns None if the stream is finished
+
+        This is an abstract method, and hence needs to be implemented by the subclasses
+        """
+        pass
+
+    def get_block(self, block_size: int) -> DataBlock:
+        """returns a block of data (of the given max size) from the stream
+
+        Args:
+            block_size (int): the size of the block of data to be returned.
+
+        Returns:
+            DataBlock:
+        """
+        # returns the next data block
+        data_list = []
+        for _ in range(block_size):
+            # get next symbol
+            s = self.get_symbol()
+            if s is None:
+                break
+            data_list.append(s)
+
+        # if data_list is empty, return None to signal the stream is over
+        if not data_list:
+            return None
+
+        return DataBlock(data_list)
+
+    @abc.abstractmethod
+    def write_symbol(self, s):
+        """writes the given symbol to the stream
+
+        The symbol can be appropriately converted to a particular format before writing.
+        This is an abstract method and so, the subclass will have to implement it
+
+        Args:
+            s (Any): symbol to be written to the stream
+        """
+        pass
+
+    def write_block(self, data_block: DataBlock):
+        """write the input block to the stream
+
+        Args:
+            data_block (DataBlock): block to be written to the stream
+        """
+        for s in data_block.data_list:
+            self.write_symbol(s)
+
+    def __enter__(self):
```

Comment:
Nice! Learnt something new here. Maybe a short comment with a link to why this is useful can be handy for those new to Python

Reply:
Thanks! done

Reply:
Thanks!